### PR TITLE
[FLINK-26383] Create model stream sink demo

### PR DESCRIPTION
**Pull request by weibozhao**
[FLINK-26383] Create model stream sink demo

### CODE REVIEW

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                tEnv.fromDataStream(env.fromCollection(validData), dataSchema).as("label, vec");
+
+        Process proc = Runtime.getRuntime().exec("ls " + tmpPath + "/data");
+        BufferedReader bufferedReader =
+                new BufferedReader(new InputStreamReader(proc.getInputStream()));
+
+        /* Loads every LogisticRegression model in model path and validates it. */
+        String modelVersion;
+        while ((modelVersion = bufferedReader.readLine()) != null) {
+            LogisticRegressionModel lrModel = new LogisticRegressionModel().setFeaturesCol("vec");
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:109]: `featuresCol` should be loaded from `metadata`.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            if (!"metadata".equals(modelVersion)) {
+                System.out.println(modelVersion);
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:111]: nit: We do not print stuff in tests.

**Review by zhipeng93 — COMMENTED** [review]
Thanks for the PR. Left some comments below.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+/** Tests online LogisticRegression model save and load. */
+public class OnlineModelSaveLoadTest {
+    @Rule public final TemporaryFolder tempFolder = new TemporaryFolder();
+    StreamTableEnvironment tEnv;
+    StreamExecutionEnvironment env;
+    Schema modelSchema = Schema.newBuilder().column("f0", DataTypes.of(DenseVector.class)).build();
+    Schema dataSchema =
+            Schema.newBuilder()
+                    .column("f0", DataTypes.of(Double.class))
```
[the remark this answers is not in the document]
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:66]: OK, I agree.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    private static final List<Row> modelData =
+            new ArrayList<>(
+                    Arrays.asList(
+                            Row.of(Vectors.dense(2.0, 4.5, 3.0)),
+                            Row.of(Vectors.dense(2.1, 4.6, 3.1)),
+                            Row.of(Vectors.dense(20.1, 5.6, 3.1)),
+                            Row.of(Vectors.dense(2.1, 4.7, 3.1))));
+
+    private static final List<Row> validData =
+            new ArrayList<>(
```
[the remark this answers is not in the document]
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:79]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        Table modelData =
+                tEnv.fromDataStream(
+                                ReadWriteUtils.loadModelData(
+                                        env,
+                                        tmpPath,
+                                        new LogisticRegressionModelData.ModelDataDecoder()))
+                        .as("label, vec");
+
+        List<Row> result = IteratorUtils.toList(tEnv.toDataStream(modelData).executeAndCollect());
+        Assert.assertEquals(result.size(), 4);
```
[the remark this answers is not in the document]
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:142]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:79]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true);
+        env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+        env.enableCheckpointing(100);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        tEnv = StreamTableEnvironment.create(env);
+        env.setParallelism(1);
+    }
+
+    @Test
+    public void saveAndLoadEverySingleModelAndValidate() throws Exception {
```
[the remark this answers is not in the document]
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:96]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                List<Row> result = IteratorUtils.toList(tEnv.toDataStream(out).executeAndCollect());
+                for (Row row : result) {
+                    Assert.assertEquals(row.getField(0), row.getField(2));
+                }
+            }
+        }
+    }
+
+    @Test
+    public void saveAndLoadAllOnlineModels() throws Exception {
```
[the remark this answers is not in the document]
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:131]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 1 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:109]: code has been removed.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 2 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:111]: remove it already.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+        env.enableCheckpointing(100);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        env.setParallelism(1);
+        tEnv = StreamTableEnvironment.create(env);
+        Schema modelSchema =
+                Schema.newBuilder().column("f0", DataTypes.of(DenseVector.class)).build();
+        Table modelDataTable = tEnv.fromDataStream(env.fromCollection(modelRows), modelSchema);
+        try {
+            tmpPath = tempFolder.newFolder().getAbsolutePath();
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:84]: As a demo, shall we include `saveModelData` as part of the two tests?

Saving/loading metadata should also be part of a demo.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                    tmpPath,
+                    new LogisticRegressionModelData.ModelDataEncoder());
+            env.execute();
+        } catch (Exception e) {
+            throw new RuntimeException("Saves model data failed!");
+        }
+    }
+
+    @Test
+    public void testSaveAndLoadSingleModel() throws Exception {
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:97]: Can we just load a specific version of the model without enumerating all of them? For example, can we just load model data of `version 2`?

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    public void testSaveAndLoadOnlineModel() throws Exception {
+        Table modelData =
+                tEnv.fromDataStream(
+                        ReadWriteUtils.loadModelData(
+                                env, tmpPath, new LogisticRegressionModelData.ModelDataDecoder()));
+
+        List<Row> loadedModelRows =
+                IteratorUtils.toList(tEnv.toDataStream(modelData).executeAndCollect());
+        int numEquals = 0;
+        for (int i = 0; i < loadedModelRows.size(); ++i) {
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:135]: Can you refine the check using `TestBaseUtils#compareResultCollections` here?

The test may still success even if the loaded model data is wrong. (e.g., four model data with version 0)

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    private static class ModelVersionAssigner<T> extends BasePathBucketAssigner<T> {
+        private long modelVersion = 0L;
+
+        @Override
+        public String getBucketId(T element, Context context) {
+            return String.valueOf(modelVersion++);
+        }
+    }
+
+    private static <T> void saveModelData(
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:155]: Can you add a java doc for this method?

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        FileSink<T> sink =
+                FileSink.forRowFormat(
+                                new org.apache.flink.core.fs.Path(path + "/data/"), modelEncoder)
+                        .withRollingPolicy(OnCheckpointRollingPolicy.build())
+                        .withBucketAssigner(new ModelVersionAssigner<>())
+                        .build();
+        model.sinkTo(sink);
+    }
+
+    private static <T> DataStream<T> loadModelData(
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:166]: ditto.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+import org.junit.rules.TemporaryFolder;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** Tests online LogisticRegression model save and load. */
+public class OnlineModelSaveLoadTest {
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:58]: Can you update the java doc or the class name here to make them consistent?

**Review by zhipeng93 — COMMENTED** [review]
Thanks for the update. This PR looks better :). Left some comments below.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 8 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:84]: Save has been move to before, the two test share the same saving, so I move it to before.

Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:84]: This demo only test modelData save and load, I think meta is not need.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 9 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:97]: I think it's all right.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 11 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:155]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 13 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:58]: OK

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 10 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:135]: OK

**Comment by weibozhao** [comment]
@zhipeng93 Thanks for your comments. I have refine the code as your comment.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 8 above)
Comment (zhipeng93) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:86]: As we are doing a demo for saving/loading a online model, I do think meta is better included.

Let's see how others think.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 9 above)
Comment (zhipeng93) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:99]: Okay. Let's see how others think~

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+                IteratorUtils.toList(tEnv.toDataStream(modelData).executeAndCollect());
+
+        TestBaseUtils.compareResultCollections(
+                loadedModelRows,
+                modelRows,
+                (o1, o2) -> {
+                    DenseVector v1 = (DenseVector) o1.getField(0);
+                    DenseVector v2 = (DenseVector) o2.getField(0);
+
+                    if (v1.values[0] > v2.values[0]) {
```
Comment (zhipeng93) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:133]: How about replace the code snippet with something like this:

`
for (int x = 0; x < min(first.length, second.length); x++) {
            int cmp = Double.compare(first[x], second[x]);
            if (cmp != 0) {
                return cmp;
            }
        }
        int cmp = first.length - second.length;
        return cmp;
`

File: flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java (hunk 8 above)
Comment (weibozhao) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineModelSaveLoadTest.java:86]: I think meta and model data saveLoad are two different thing. Here we just give a demo for model data save and load.

**Comment by zhipeng93** [comment]
Closing for now. We can open it later if needed.

**Pull request closed without merging**